File: scripts/check_i18n.js
// Simple approach: use regex to extract all key paths from i18n.js
const fs = require('fs');
const path = require('path');

const i18nPath = path.join(__dirname, '..', 'web', 'src', 'i18n.js');
const content = fs.readFileSync(i18nPath, 'utf-8');

// Strategy: Use a line-by-line approach to build key paths based on indentation
function extractKeysByIndentation(text) {
  const keys = new Set();
  const lines = text.split('\n');
  const stack = []; // {indent, key}
  
  for (const line of lines) {
    // Match patterns like "key: 'value'," or "key: {" or "key: number,"
    const m = line.match(/^(\s*)(\w+)\s*:\s*(.*)$/);
    if (!m) continue;
    
    const indent = m[1].length;
    const key = m[2];
    const value = m[3].trim();
    
    // Pop stack until we find the parent (indent < current indent)
    while (stack.length > 0 && stack[stack.length - 1].indent >= indent) {
      stack.pop();
    }
    
    const fullKey = stack.length > 0 
      ? stack.map(s => s.key).join('.') + '.' + key 
      : key;
    
    if (value.startsWith('{')) {
      // Nested object - push to stack
      stack.push({ indent, key });
    } else {
      // Leaf value (string, number, boolean, template literal)
      keys.add(fullKey);
    }
  }
  
  return keys;
}

const allKeys = extractKeysByIndentation(content);

// Now filter to get zh and en keys
const zhKeys = new Set([...allKeys].filter(k => k.startsWith('zh.')));
const enKeys = new Set([...allKeys].filter(k => k.startsWith('en.')));

// Strip the 'zh.' and 'en.' prefix
const zhKeysStripped = new Set([...zhKeys].map(k => k.replace(/^zh\./, '')));
const enKeysStripped = new Set([...enKeys].map(k => k.replace(/^en\./, '')));
const allDefinedKeys = new Set([...zhKeysStripped, ...enKeysStripped]);

// Extract all t('key') calls from Vue/JS files
const usedKeys = new Set();
const srcDir = path.join(__dirname, '..', 'web', 'src');
function walkDir(dir) {
  const files = fs.readdirSync(dir);
  for (const f of files) {
    const fullPath = path.join(dir, f);
    const stat = fs.statSync(fullPath);
    if (stat.isDirectory()) {
      walkDir(fullPath);
    } else if (f.endsWith('.vue') || f.endsWith('.js')) {
      if (f === 'i18n.js') continue;
      const c = fs.readFileSync(fullPath, 'utf-8');
      const matches = c.matchAll(/t\(\s*['"]([a-zA-Z][a-zA-Z0-9_]*\.[a-zA-Z][a-zA-Z0-9_.]*)['"]\s*[,)]/g);
      for (const m of matches) {
        usedKeys.add(m[1]);
      }
    }
  }
}
walkDir(srcDir);

// Find missing keys
const missing = [...usedKeys].filter(k => !allDefinedKeys.has(k)).sort();
if (missing.length > 0) {
  console.log(`=== ${missing.length} MISSING i18n keys (used but not defined) ===`);
  for (const k of missing) {
    const inZh = zhKeysStripped.has(k);
    const inEn = enKeysStripped.has(k);
    console.log(`  ${k}  [zh: ${inZh ? 'YES' : 'NO'}, en: ${inEn ? 'YES' : 'NO'}]`);
  }
} else {
  console.log('All used i18n keys are defined!');
}

// Check zh-en symmetry
const zhOnly = [...zhKeysStripped].filter(k => !enKeysStripped.has(k)).sort();
const enOnly = [...enKeysStripped].filter(k => !zhKeysStripped.has(k)).sort();

if (zhOnly.length > 0) {
  console.log(`\n=== ${zhOnly.length} keys in zh but MISSING in en ===`);
  for (const k of zhOnly) {
    console.log(`  ${k}`);
  }
}
if (enOnly.length > 0) {
  console.log(`\n=== ${enOnly.length} keys in en but MISSING in zh ===`);
  for (const k of enOnly) {
    console.log(`  ${k}`);
  }
}

console.log(`\n=== Summary ===`);
console.log(`Used keys: ${usedKeys.size}`);
console.log(`Defined zh keys: ${zhKeysStripped.size}`);
console.log(`Defined en keys: ${enKeysStripped.size}`);
console.log(`Missing (used but not defined): ${missing.length}`);
console.log(`zh-only (missing in en): ${zhOnly.length}`);
console.log(`en-only (missing in zh): ${enOnly.length}`);

File: scripts/check_i18n.py
"""Check for missing i18n keys - only check dotted keys like 'devices.address' that look like real i18n keys."""
import re
import os
import sys

# Extract all t('key') calls from Vue/JS files
used_keys = set()
src_dir = os.path.join(os.path.dirname(__file__), '..', 'web', 'src')
for root, dirs, files in os.walk(src_dir):
    for f in files:
        if f.endswith(('.vue', '.js')) and f != 'i18n.js':
            filepath = os.path.join(root, f)
            with open(filepath, encoding='utf-8') as fh:
                content = fh.read()
            # Match t('some.key') patterns - only dotted keys with letters
            matches = re.findall(r"t\(\s*['\"]([a-zA-Z][a-zA-Z0-9_]*\.[a-zA-Z][a-zA-Z0-9_.]*)['\"]\s*[,)]", content)
            for m in matches:
                used_keys.add(m)

# Load i18n.js and parse the nested structure to get all valid key paths
i18n_path = os.path.join(src_dir, 'i18n.js')
with open(i18n_path, encoding='utf-8') as fh:
    i18n_content = fh.read()

def extract_keys_from_obj(content, prefix=''):
    """Extract all dotted key paths from a JS object literal."""
    keys = set()
    i = 0
    while i < len(content):
        m = re.search(r"^\s*(\w+)\s*:", content[i:], re.MULTILINE)
        if not m:
            break
        key_name = m.group(1)
        value_start = i + m.end()
        j = value_start
        while j < len(content) and content[j] in ' \t':
            j += 1
        if j >= len(content):
            break
        full_key = f"{prefix}.{key_name}" if prefix else key_name
        if content[j] == '{':
            depth = 1
            k = j + 1
            while k < len(content) and depth > 0:
                if content[k] == '{':
                    depth += 1
                elif content[k] == '}':
                    depth -= 1
                k += 1
            nested_content = content[j+1:k-1]
            nested_keys = extract_keys_from_obj(nested_content, full_key)
            keys.update(nested_keys)
            i = k
        elif content[j] in "'\"":
            quote = content[j]
            k = j + 1
            while k < len(content) and content[k] != quote:
                if content[k] == '\\':
                    k += 1
                k += 1
            keys.add(full_key)
            i = k + 1
        else:
            k = j
            while k < len(content) and content[k] not in ',\n}':
                k += 1
            keys.add(full_key)
            i = k
    return keys

defined_keys = extract_keys_from_obj(i18n_content)

# Find missing keys
missing = sorted(used_keys - defined_keys)
if missing:
    print(f"=== {len(missing)} MISSING i18n keys (used but not defined) ===")
    for k in missing:
        print(f"  {k}")
else:
    print("All used i18n keys are defined!")

print(f"\n=== Summary ===")
print(f"Used dotted keys: {len(used_keys)}")
print(f"Defined keys: {len(defined_keys)}")
print(f"Missing: {len(missing)}")
